foursight_core/react/api/datetime_utils.py

```diff
@@ -48,23 +48,73 @@ def convert_time_t_to_useastern_datetime_string(time_t: int) -> str:
     except Exception:
         return "datetime-parse-error"
 
+
 def convert_time_t_to_datetime(time_t: int) -> datetime.datetime:
     """
     Converts the given "epoch" time (seconds since 1970-01-01T00:00:00Z)
     to a datetime and returns its value.
     """
     return datetime.datetime.fromtimestamp(time_t, tz=pytz.UTC)
 
+
 def convert_datetime_to_time_t(value: datetime.datetime) -> int:
     """
     Converts the given datatime value to an "epoch"
     time (seconds since 1970-01-01T00:00:00Z) and returns its value.
     """
     return int(value.timestamp())
 
+
 def convert_utc_datetime_to_cookie_expires_format(t) -> str:
     """
     Converts the given UTC datetime object or string format suitable
     for use by a cookie expires date/time value.
     """
     return t.strftime("%a, %d %b %Y %H:%M:%S GMT")
+
+
+def convert_uptime_to_datetime(uptime: str, relative_to: datetime = None) -> Optional[datetime.datetime]:
```

Comment:
Might also work to just do (lightly tested):

```
def parse_relative_time_string(s):
    parts = [x for x in s.split(' ') if x is not '']
    if len(parts) % 2 != 0:
        raise ValueError(f"Relative time strings are an even number of tokens"
                         f" of the form '<n1> <unit1> <n2> <unit2>...': {s!r}")
    kwargs = {}
    for i in range(len(parts) // 2):
        # Canonicalize "1 week" or "1 weeks" to "weeks": 1.0 for inclusion as kwarg to to timedelta
        # Uses specialized knowledge that all time units don't end in "s" but pluralize with "+s"
        kwargs[parts[i+1].rstrip('s')+"s"] = float(parts[i])
    return datetime.timedelta(**kwargs)
```

We should probably have something like this in `dcicutils.lang_utils` as an inverse of `relative_time_string`.

```
>>> relative_time_string(parse_relative_time_string("1 hour"))
'1 hour'
>>> relative_time_string(parse_relative_time_string("1.5 hours"))
'1 hour, 30 minutes'
```

I would then maybe do (untested):

```
def convert_uptime_to_datetime(uptime: str, relative_to: datetime = None) -> Optional[datetime.datetime]:
    if not uptime:
        return None
    uptime_delta = parse_relative_time_string(uptime)
    start_time = datetime.datetime.now() - uptime_delta()
    return start_time
```

This would simplify the part where you're dependent on now in the current timezone or now in UTC by whether you used `.now()` or `.utcnow()`. You could perhaps just start with .utcnow() and attach a timezone of `pytz.UTC` and then adjust for local time with `.astimezone` once you have a proper foothold.
Doing:

```
>>> datetime.datetime.now()
datetime.datetime(2022, 11, 16, 15, 34, 8, 753294)
```

may seem to get you what you want but you won't know what the local timezone is. If instead you do:

```
>>> str(datetime.datetime.utcnow().replace(tzinfo=pytz.UTC).astimezone(pytz.timezone("US/Eastern")))
'2022-11-16 15:34:13.905371-05:00'
```

then you can be sure you are starting from a known place and that you've made appropriate adjustments for a given named timezone that we could declare somewhere (perhaps in the GAC even).

Reply:
This implementation has bugs, but I have a better one I'll put up for review as part of dcicutils.

Reply:
[dcicutils PR #228](https://github.com/4dn-dcic/utils/pull/228) has the corrected definition.